array_methods_assignment_solution.js

```diff
@@ -0,0 +1,404 @@
+    return acc + currentEl;
+  });
+  return newArr;
+}
+console.log(sumAll([-11,2,3,4]));
+
+
+
+
+
+// 12. Use `myMap` to build a new array that doubles each element. Then chain `myReduce` to find the product of all the elements. 
+console.log("######### Prob 12 #########");
+
+const doubledProduct = (arr) => {
+  let myNewArr = arr.myMap(el => {
+    return el*2;
+  }).myReduce((acc, curEl) => {
+    return acc + curEl
+  })
+  return myNewArr;
+}
+
+console.log(doubledProduct([1,2,3]))
+
+
+
+
+
+// 13. Use `myReduce` to find the largest number in an array. 
+console.log("######### Prob 13 #########");
+
+const largestEl = (arr) => {
+  let newArr = arr.myReduce((acc, curEl) => {
+    if(acc > curEl){
+      return acc;
+    }else{
+      return curEl;
+    }
+  });
+  return newArr;
+}
+console.log(largestEl([1,2,3,4,,12,4]))
+
+
+
+
+
+// 14. Write a function `elementDivisibleBy` with parameters `divisor` and `arr`.
+// Use `myFilter` to return a new array of every element of arr that can be evenly divided by divisor.
+console.log("######### Prob 14a #########");
+
+const elementDivisibleBy = (arr, divisor) => {
+  let newArr = arr.myFilter(el => {
+    return el%divisor===0;
+  });
+  return newArr;
+}
+console.log(elementDivisibleBy([1,2,3,4,5,6], 2));
+
+
+
+
+
+
+// 14. Write a `myJoin` function. 
+console.log("######### Prob 14b #########");
+
+Array.prototype.myJoin = function(joiningChar){
+  let str = '';
+  this.forEach(el => {
+    str += el + joiningChar;
+  });
+  return str;
+}
+
+let arr7 = ['a', 'b', 'c', 'd']
+let newArr7 = arr7.myJoin("");
+console.log(newArr7);
+
+
+
+
+
+// 15. Write a `mySlice` function.
+console.log("######### Prob 15 #########");
+Array.prototype.mySlice = function(start, end){
+let arr = [];
+  
+  for(let i = start; i < end; i++){
+    arr.push(this[i]);
+  }
+  return arr;
+}
+
+console.log([1, 3, 4, 5].mySlice(1, 4))
+
+
+// 16. Write a function `countZeroes`, which takes an array of numbers as 
+//its argument and returns the amount of zeroes that occur in it.
+// Use reduce.
+console.log("######### Prob 16 #########");
+const countZeroes = (arr) => {
+  let count = 0;
+  arr.reduce((acc, curEl) => {
+    if(curEl === 0 || acc === 0){
+       count++;    
+     }
+  });
+  return count;
+}
+
+console.log(countZeroes([0,1,2,3,0,0,43,0]));
+
+
+
+
+// 17. Write a `numberTimesIdx` that uses `map` and multiples each number in the array by it's index. 
+console.log("######### Prob 17 #########");
+const numberTimesIdx = (arr) => {
+  let newArr = arr.map((el, index) => {
+    return el * index;
+  });
+  return newArr;
+}
+
+console.log(numberTimesIdx([1,2,3,4]));
+
+
+
+// Bonus: Write a `myFlatten`. This should take a multi-dimensional array and return it as one array. 
+// ```js
+// let arr = [1, 2, [3, 4, 5, [6, 7, 8]]]
+//   arr.myFlatten();
+//   // => [1, 2, 3, 4, 5, 6, 7, 8]
+// ```
+
+console.log("######### BONUS #########");
+const myFlatten = (arr) => {
```

Comment:
will work fro 3 dementions but not 4. This is why we'll need to use recursion. We haven't taught that yet, so great attempt!